### ISSUE TITLE
Having duplicate style rules embeded in one element

### ISSUE THREAD

**Issue by PikachuEXE** [Mange/roadie#111]
# Env
- `roadie` `3.1.0`
- `roadie-rails` `1.1.0`
- `rails` `4.1.14`

# Example
This is copied from the source shown by Rails mailer preview.
```html
<a class="btn btn-show" href="(The URL)" style="cursor:pointer;text-decoration:none;color:#00a0e1;-webkit-transition:color 0.2s linear, background-color 0.2s linear;-moz-transition:color 0.2s linear, background-color 0.2s linear;transition:color 0.2s linear, background-color 0.2s linear;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;color:#333333;text-decoration:none;display:inline-block;padding:4px 12px;margin-bottom:0;font-size:13px;line-height:20px;text-align:center;vertical-align:middle;cursor:pointer;-webkit-border-radius:3px;-moz-border-radius:3px;border-radius:3px;-moz-box-shadow:none;-webkit-box-shadow:none;box-shadow:none;-webkit-transition:background-color 0.2s linear;-moz-transition:background-color 0.2s linear;transition:background-color 0.2s linear;border:1px solid #cccccc;background-color:#ffffff;color:#000000;text-decoration:none;background-image:none;background-position:0;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;width:100%;padding-left:0;padding-right:0;font-size:14px;font-weight:bold;box-sizing:content-box !important;text-shadow:none !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important;border:0 !important;background-color:#00a0e1 !important;color:#ffffff !important">Link Text</a>
```

**Comment by Mange** [comment]
I can confirm this problem with this example below:

```ruby
  it "does not inline the same properties several times" do
    document = Roadie::Document.new <<-HTML
      <head>
        <link rel="stylesheet" href="hello.css">
      </head>
      <body>
        <p class="hello world">Hello world</p>
      </body>
    HTML

    document.asset_providers = TestProvider.new("hello.css" => <<-CSS)
      p { color: red; }
      .hello { color: red; }
      .world { color: red; }
    CSS

    result = parse_html document.transform
    expect(result).to have_styling(['color', 'red']).at_selector('p')
  end
```

```
Failures:

  1) Roadie functionality does not inline the same properties several times
     Failure/Error: expect(result).to have_styling(['color', 'red']).at_selector('p')
       expected styles at "p" to be:
       ["color", "red"]
       but was:
       [["color", "red"], ["color", "red"], ["color", "red"]]
     # ./spec/integration_spec.rb:196:in `block (2 levels) in <top (required)>'
```

I'll fix this right now.

**Comment by Mange** [comment]
I just realized when opening up `Inliner` specs that this is by design, somewhat. Roadie does not filter out duplicate rules in order to support styles like this:

```css
.heading {
  background-color: #eee;
  background-color: rgba(255, 255, 255, 0.8);
}
```

(e.g. that if a client does not support the next style, the previous one will be used instead.)

I might still be able to handle this a bit smarter. Consider this: If two style are identical, the previous ones *could* be removed: `color: red; color: blue; color: red` → `color: blue; color: red;`.

I think this could break in some cases, but I have a hard time coming up with examples that would break.

We have shorthands that could interfere with the other styles...

```css
p {
  border-top-color: red;
  border: 1px solid black;
  border-top-color: red;
}

p {
  border: 1px solid black;
  border-top-color: red;
  border: 1px solid black;
}
```

But as you can see in both these cases, it doesn't matter. Either the styles were overwritten anyway, or they will not be overwritten anyway.

I will do some experiments with this and see how it works out.

**Comment by Mange** [comment]
@PikachuEXE I've just published a new branch called `deduplication`. Could you see if this branch works better for you? It shouldn't be able to completely eliminate the problem, but I hope that it would have a pretty significant reduction in size for you.

It goes from 121 properties to 38 of them when I test it with your example, just 31% of the original number of properties.

**Comment by PikachuEXE** [comment]
Yes the branch does reduce the no. of embedded style rules.
I notice this bug when testing email on staging, seeing the email is cut off by GMail.

**Comment by PikachuEXE** [comment]
Any more changes required for this to be merged into `master` ?
Or is there anything I can provide?

**Comment by jng5** [comment]
Honestly, it looks like @PikachuEXE needs to clean up the CSS before it's inlined. When I see so many duplicate definitions, I smell a framework…

I am wary that deferring from @Mange's original implementation of allowing multiple definitions (by design) would open up possibilities for bugs from changing code as intended. Code as you intend (don't define it multiple times to begin with) and you should be fine.

***IF*** this is added, it should be optional.

**Comment by PikachuEXE** [comment]
I agree it should be optional.

**Comment by Mange** [comment]
Julie, I agree.

I will make it default on because I still cannot come up with a case that
would break as long as only identical entries are de-duplicated. If it
turns out I was wrong, the setting will be a welcome workaround and I will
then switch the default if the case is common enough.

What say thee?

tors 3 dec. 2015 10:41 PikachuEXE <notifications@github.com> skrev:

**Comment by jng5** [comment]
Default on? 

I would make it default off, that is, current behavior is remains the same. Your specs show the most common use case where you want something duplicated the `background` property. I also recently had a case where to support Outlook 2000/02/03, I also needed `display:inline; display:inline-block;` on a `<div>`.

It would be frustrating if suddenly that behavior changed on an update. I would consider this a breaking change…

**Comment by PikachuEXE** [comment]
@jng5 If you look at the spec in https://github.com/Mange/roadie/commit/49e22e63410e76f656ea0b3238232de579a23aee
The code would not change anything in your example

**Comment by jng5** [comment]
No it does not change the spec. But I still think it's adding something just to fix *someone else's bad code* - which doesn't sit right with me.

But it's @Mange's decision, since he's built and supports this awesome gem :)

@PikachuEXE - I used to have similar problems. But it came from using a framework or my own bad habits from CSS for the web. Then I realized I needed to be more explicit for email and not relying on cascades. I urge you to consider examining the root of the problem. That example you gave looked straightforward. But if you were ever to do more complicated emails, I imagine your code would be a nightmare to debug…

**Comment by Mange** [comment]
Good tips, over all. It's important to use simpler styles for emails, for
sure.

I wanted something like this myself before, as I like optimizing stuff and
compressing the final HTML output is something I think Roadie should strive
for. This issue was mostly a good case and motivation for me to work on it,
and even though most cases will never be this extreme, I can see how a lot
of small repetitions can end up in emails and I'm glad if Roadie can help
in these cases. After all, the idea is to make email less painful for Web
developers. :-)

mån 7 dec. 2015 09:41 Julie Ng <notifications@github.com> skrev:

**Comment by jng5** [comment]
Well said :) I think you're right and I need to just let go of my high standards for myself. Even I end up with sloppy code. 

I have yet to read the changes in the branch, but I wrote something similar for my workflow. I try to work with reusable components. But sometimes there are exceptions to a rule and that's when I get similar duplicate CSS issues, which I then output as a semi-warning to the console when inlining. But it's just one duplicate property, not 83 duplicate properties in the example above.

I didn't consider the motivation part. That's more important than my arguments :)

**Comment by Mange** [comment]
I just released version 3.1.1 with this in. I decided not to add a configuration option anyway as I'd need to replicate it all out into `roadie-rails` for most people, and then maintain the option over time (requiring a major version bump to remove it in the future). I'd rather have people open an issue here if there's any problems, and a valid work-around is to lock the version to 3.1.0.
